Add a send_request function to NetworkService

eskimor opened the pull request:
This function differs from the already provided `request` method in two ways.

1. It allows sending of messages to currently disconnected peers, which is required for my first go at request based availability distribution.
2. Instead of creating a sender/receiver pair and awaiting the receiver, it takes a sender as argument. This suits our use in Polkadot better, where the subsystem will hold the corresponding receiver.


This PR changes the signature of some apparently only internally used `send_request` functions to accept an additional boolean parameter.

bkchr commented:
>     2\. Instead of creating a sender/receiver pair and awaiting the receiver, it takes a sender as argument. This suits our use in Polkadot better, where the subsystem will hold the corresponding receiver.

The `request` function is returning a `Future`. So, you have exactly the same interface as with your `oneshot::Receiver`. You have a future that you can await on.

eskimor commented:
> The `request` function is returning a `Future`. So, you have exactly the same interface as with your `oneshot::Receiver`. You have a future that you can await on.

That's not true. For `response` - I get a future, which I would have to thread back to the sending subsystem. For `send_request` that's not necessary, as the sending subsystem already has the corresponding future.

bkchr commented:
> Assuming the end user of the API would be the function caller this would of course be correct, but that is not the case in Polkadot, where subsystems communicate with the network bridge via message passing.

The polkadot subsystem can just send its oneshot::Sender and you do `async move { sender.send(future.await); }`.

eskimor commented:
> The polkadot subsystem can just send its oneshot::Sender and you do `async move { sender.send(future.await); }`

That's true, but it felt stupid to create another sender/receiver pair when there is one already.

tomaka commented:
I'd strongly prefer to have the same API as `request`.
Adjusting the API in Substrate to the implementation details of Polkadot feels wrong.
Passing a sender also makes it impossible to implement cancelling requests if the future is dropped.

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+		protocol: impl Into<Cow<'static, str>>,
+		request: Vec<u8>,
+		tx: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+		connect: bool,
```
```suggestion
```

In my opinion, if the user wants `connect` to be false, they should call `request`. If they want `connect` to be true, they should call `send_request`.

eskimor replied:
If connect/not connect semantics end up being the only difference, then for sure. If not, I think having that simple use case function `request` and the power horse `send_request` with full flexibility makes sense. I don't think, that if I preferred a channel based API, that I also always want 'auto connect", as it has downsides as you mentioned.

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	/// Apart from exposing whether or not we want to connect, in case of a disconnected peer, this
+	/// function also takes a sender for sending the response back, instead of returning the
+	/// response directly.
+	pub fn send_request(
```
Assuming we remove the `connect` parameter, a better name would be preferable.

```suggestion
	pub fn connect_and_request(
```

eskimor commented:
Hmm, not sure I can agree. The request code path is already quite hard to follow, with quite a few layers of indirection. Not exposing the API that is already there and is needed/preferred in Polkadot, just adds another layer. Given the fact that it is implemented internally as a sender/receiver pair also suggests that it is not an esoteric use case in Polkadot to prefer such an API, but rather something that is likely to come up often, as we already depend on such an API both in Substrate and in Polkadot.

>     * Passing a sender also makes it impossible to implement cancelling requests if the future is dropped.

Why? There is (poll_canceled)[https://docs.rs/futures/0.3.12/futures/channel/oneshot/struct.Sender.html#method.poll_canceled] to do precisely that. At least if I understood your concern correctly.

>     * The receiving side of a channel can produce a "cancelled" error if the sender has been closed, which needs to be handled properly and is less explicit than `RequestFailure`.

That is true, but obviously already handled on the Polkadot side. I agree, that this is less explicit than a constructor in `RequestFailure`, but on the other hand getting a cancelled error, it is still pretty clear what happened.

eskimor commented:
Adding to this, I would see the `request` function as a good and easy default  for simple use cases and the `send_request` function as a more flexible/more bare-bone interface for more advanced use cases. I also did not get the feeling, `send_request` would expose some rather hidden implementation detail or something, although I am sensing you are suggesting something in that direction? What are the reasons, you think exposing a Sender/Receiver interface is not such a good idea? Using a <dyn Future> is sure more generic, but I am not sure why we would want/need that level of generality here. In my view a oneshot::Sender/Receiver is a pretty good standard infra to build upon.

eskimor commented:
> I would encourage you to use a `Pin<Box<dyn Future<Output = ...>>>` in Polkadot to inform of the response, instead of a `Receiver<...>`.

If that is about hiding implementation details and getting the nicer API, that all errors are contained in `RequestFailure`, wrapping the `Receiver` in a `IncomingRequestReceiver` struct might even be nicer and also do the trick? Although I am not sure, wrapping the oneshot::Receiver API just for that is worth it.

tomaka commented:
Yes my grief is about exposing implementation details. I'd rather not tie the networking API to channels.
At the moment, there is one, for receiving incoming requests, not because we're using a channel internally (we're not) but because I felt that it would be easy from a user perspective to understand an API where you plug a channel sender, while any alternative would have been way more complicated.
There's no complexity problem for `request`. It's simple to use.

> but I am not sure why we would want/need that level of generality here

What's wrong with `Pin<Box<dyn Future>>`? What is the reason why we wouldn't want it?

eskimor commented:
> Yes my grief is about exposing implementation details. I'd rather not tie the networking API to channels.
> At the moment, there is one, for receiving incoming requests, not because we're using a channel internally (we're not) but because I felt that it would be easy from a user perspective to understand an API where you plug a channel sender, while any alternative would have been way more complicated.

I am not sure we should be seeing the channels as an implementation detail. As you mentioned, we already have a channel based interface for receiving incoming requests. It makes sense to me to offer such an interface also for sending requests (at least optionally, as suggested with `send_request`). It is also not an uncommon interface for a future based API, is quite nice for message passing and also it is already there, we are just hiding it.

Do we have plans on getting rid of the channel internally? Exposing a channel based API even in that case, would not be an issue, but I agree that if that is the case we should have a solid case of exposing such an API to be a good user interface, as then we would end up just making it up for the user facing code.

> There's no complexity problem for `request`. It's simple to use.

For simple use cases, but not as nice for communication via message passing which is not uncommon in multithreaded environments (Polkadot and Substrate being good examples). It is of course not a terrible issue to create a channel based interface on top, but I would also rather not needlessly complicate things more.


> What's wrong with `Pin<Box<dyn Future>>`? What is the reason why we wouldn't want it?

It feels a bit clunky and needlessly dynamic, in particular I am not sure I understand why I should prefer it over a Receiver. But I guess, that boils down to whether we want channels in the public API or not.


>     * Adjusting the API in Substrate to the implementation details of Polkadot feels wrong.

I am also not sure about this. If Polkadot was somehow esoteric with patterns no-one other should ever use, then yes. But if it is just a user of our API, then its  needs should absolutely inform our API design! Of course we should also not delve into featuritis, that would not help either :-)

tomaka commented:
I looked at your Polkadot PR, and I now realize that you're creating the channel ahead of time in a total different part of the code.
I still don't think this API makes sense when taken in isolation, but I can get behind the argument that "changing Polkadot would require either a big refactoring or introducing code that's a little stupid".

I'll do another review.

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+		protocol: impl Into<Cow<'static, str>>,
+		request: Vec<u8>,
+		tx: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+		connect: bool,
```
Instead of a `bool`, I'd suggest a new `IfDisconnectedBehaviour` enum, to make it clear what this does.
```suggestion
		connect: IfDisconnectedBehaviour,
```

```rust
pub enum IfDisconnectedBehaviour {
    TryConnect,
    ImmediateError,
}
```

You don't need to pass it around, it's just to make the public API clear. Feel free to pass a `bool` to the internal methods.

eskimor replied:
I like this, was already tempted to do something like that. I would not mind changing the internal functions as well, I think this would improve readability over all.

eskimor replied:
Ok, this would get noisy, due to the long name :-P Will probably stick with the public API for now.

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	/// Apart from exposing whether or not we want to connect, in case of a disconnected peer, this
+	/// function also takes a sender for sending the response back, instead of returning the
+	/// response directly.
+	pub fn send_request(
```
```suggestion
	pub fn request_channel(
```

I really don't like `send_request` because it doesn't mean much :grimacing: and in particular doesn't show any difference in what it does compared to `request`.

eskimor replied:
I took inspiration [here](https://github.com/paritytech/substrate/blob/b2b9af4f1553fd82c98a816486e9ed0e117104e5/client/network/src/request_responses.rs#L274) for the name, as it has the exact same interface :-) 

Not sure about `request_channel` though, as it sounds like requesting a channel or alternatively like a request channel, when in fact we have a channel for responses. I will try to come up with something.

tomaka commented on client/network/src/service.rs:
```diff
@@ -828,12 +828,8 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		request: Vec<u8>
 	) -> Result<Vec<u8>, RequestFailure> {
```
```suggestion
		request: Vec<u8>,
		connect: IfDisconnectedBehaviour
	) -> Result<Vec<u8>, RequestFailure> {
```

Assuming that the only difference between `request` and `send_request`/other-name would be the channel, I'd suggest to add `IfDisconnectedBehaviour` here as well.

eskimor replied:
This would be a breaking change though, I guess I should this would then be release notes worthy?

tomaka replied:
This is done with a `B5` label

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		}
 	}
 
+	/// Variation of `request` which can connect to peers when necessary.
+	///
+	/// In some cases we do request data, without prior notification. For those cases this function
+	/// exists. It has no guarantees on connection staying alive, so you might still want to pass
+	/// false to the `connect` parameter and ensure liveness of connections via peer sets.
+	/// Connecting to a peer can take up to several seconds, if a peer exposes multiple
+	/// addresses, but is only reachable by some of them, for example.
+	///
+	/// Apart from exposing whether or not we want to connect, in case of a disconnected peer, this
+	/// function also takes a sender for sending the response back, instead of returning the
+	/// response directly.
```
Assuming you follow my other suggestions, this would need an update.

tomaka reviewed (changes requested):
I think it would make sense if the only difference between `request` and `send_request`/other-name is that the latter sends the response on the channel.

eskimor commented:
Ok, I introduced the `IfDisconnectedBehaviour` and updated the docs. I ended up introducing `IfDisconnectedBehaviour` also for the internal function calls as it made the code more readable and straight forward.

I did shorten the name though from `IfDisconnectedBehaviour` to just `IfDisconnected`. Reason being, I introduced that enum right between other `Behaviour`s, which are behaviours in the libp2p sense, so that name for the enum seemed slightly confusing. Also the shorter name makes function calls a bit more readable as well: IfDisconnected::TryConnect ... reads almost like an English sentence describing what happens.

`request` now also takes that additional parameter.

tomaka commented on client/network/src/service.rs:
```diff
@@ -844,6 +842,32 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+		request: Vec<u8>,
+		tx: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+		connect: IfDisconnected,
+		) {
```
```suggestion
	) {
```

eskimor replied:
damn it, thanks!

romanb commented on client/network/src/service.rs:
```diff
@@ -844,6 +842,32 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	/// integrate with message passing APIs.
+	///
+	/// Keep in mind that the connected receiver might receive a `Canceled` event in case of a
+	/// closing connection, this is expected behaviour. In `request` you would get a
```
```suggestion
	/// closing connection. This is expected behaviour. With `request` you would get a
```

romanb commented on client/network/src/service.rs:
```diff
@@ -844,6 +842,32 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	///
+	/// Keep in mind that the connected receiver might receive a `Canceled` event in case of a
+	/// closing connection, this is expected behaviour. In `request` you would get a
+	/// `RequestFailure::Network(OutboundFailure::ConnectionClosed` in that case.
```
```suggestion
	/// `RequestFailure::Network(OutboundFailure::ConnectionClosed)` in that case.
```

romanb commented on client/network/src/service.rs:
```diff
@@ -844,6 +842,32 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
+	/// Keep in mind that the connected receiver might receive a `Canceled` event in case of a
+	/// closing connection, this is expected behaviour. In `request` you would get a
+	/// `RequestFailure::Network(OutboundFailure::ConnectionClosed` in that case.
+	pub fn detached_request(
```
```suggestion
	pub fn start_request(
```

eskimor replied:
I like that one as well. What do you say @tomaka ?

tomaka replied:
I prefer `detached` but I don't care so much

eskimor replied:
Ok then, I will go with `start_request` as it reminds of some other API with similar semantics.

romanb commented on client/network/src/service.rs:
```diff
@@ -844,6 +842,32 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		}
 	}
 
+	/// Variation of `request` which takes a sender for sending responses back.
```
```suggestion
	/// Variation of `request` which starts a request whose response is delivered on a provided channel.
```

romanb reviewed (approved):
Generally looks ok to me, I just left some small suggestions for naming and API comments that you can choose to ignore if you don't consider them an improvement. However, if I understand correctly, this relies on the automatic dialing of `libp2p-request-response`. Note that substrate's `RequestResponsesBehaviour` [still considers it an error if an inner `RequestResponseBehaviour` emits a `DialPeer` command](https://github.com/paritytech/substrate/blob/86faec7b2557ede6a3ccacb71138b6b6f5b6e6c6/client/network/src/request_responses.rs#L511).  This error logging should probably be removed then to avoid confusion.

eskimor commented:
Thanks a lot @romanb! Lots of good catches!

eskimor commented:
Why does the github-actions keep adding 'A7-needspolkadotpr`, when this is clearly not the case? Because I referenced this PR from a Polkadot PR?
